### PR TITLE
ARROW-8377: [CI][C++][R] Build and run C++ tests on Rtools build

### PR DESCRIPTION
**Pull request by nealrichardson**

### CODE REVIEW

**Comment by nealrichardson** [comment]
This is failing on the googletest_ep:

```
[  2%] Performing build step for 'googletest_ep'
CMake Error at D:/a/arrow/arrow/src/build-x86_64-cpp/googletest_ep-prefix/src/googletest_ep-stamp/googletest_ep-build-DEBUG.cmake:50 (message):
-- stdout output is:
  Command failed: 2
make[3]: Entering directory '/d/a/arrow/arrow/src/build-x86_64-cpp/googletest_ep-prefix/src/googletest_ep-build'

make[4]: Entering directory '/d/a/arrow/arrow/src/build-x86_64-cpp/googletest_ep-prefix/src/googletest_ep-build'
   'make'
make[5]: Entering directory '/d/a/arrow/arrow/src/build-x86_64-cpp/googletest_ep-prefix/src/googletest_ep-build'

Scanning dependencies of target gtest
  See also
make[5]: Leaving directory '/d/a/arrow/arrow/src/build-x86_64-cpp/googletest_ep-prefix/src/googletest_ep-build'

make[5]: Entering directory '/d/a/arrow/arrow/src/build-x86_64-cpp/googletest_ep-prefix/src/googletest_ep-build'
    D:/a/arrow/arrow/src/build-x86_64-cpp/googletest_ep-prefix/src/googletest_ep-stamp/googletest_ep-build-*.log
[ 12%] Building CXX object googlemock/gtest/CMakeFiles/gtest.dir/src/gtest-all.cc.obj
make[5]: Leaving directory '/d/a/arrow/arrow/src/build-x86_64-cpp/googletest_ep-prefix/src/googletest_ep-build'

make[4]: Leaving directory '/d/a/arrow/arrow/src/build-x86_64-cpp/googletest_ep-prefix/src/googletest_ep-build'
CMake Error at D:/a/arrow/arrow/src/build-x86_64-cpp/googletest_ep-prefix/src/googletest_ep-stamp/googletest_ep-build-DEBUG.cmake:60 (message):
make[3]: Leaving directory '/d/a/arrow/arrow/src/build-x86_64-cpp/googletest_ep-prefix/src/googletest_ep-build'

-- stderr output is:
make[3]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
  Stopping after outputting logs.


D:/a/arrow/arrow/src/build-x86_64-cpp/googletest_ep-prefix/src/googletest_ep/googletest/src/gtest-all.cc:1:0: error: -fPIC ignored for target (all code is position independent) [-Werror]
 // Copyright 2008, Google Inc.
 ^
cc1plus.exe: all warnings being treated as errors
make[5]: *** [googlemock/gtest/CMakeFiles/gtest.dir/build.make:63: googlemock/gtest/CMakeFiles/gtest.dir/src/gtest-all.cc.obj] Error 1
make[4]: *** [CMakeFiles/Makefile2:215: googlemock/gtest/CMakeFiles/gtest.dir/all] Error 2
```

**Comment by wesm** [comment]
Appears we aren't supposed to use `-fPIC` with MinGW

**Comment by wesm** [comment]
Have we ever built all the unit tests with MinGW? It seems perhaps not

**Comment by nealrichardson** [comment]
We do here, right? https://github.com/apache/arrow/blob/master/.github/workflows/ruby.yml#L141

**Comment by nealrichardson** [comment]
Maybe that's only 64 bit though? Anyway I thought we did on Appveyor. Surely @kou knows.

**Comment by wesm** [comment]
@nealrichardson no, see that `ARROW_BUILD_TESTS` is OFF in that entry

**Comment by wesm** [comment]
Now there's a weird make error =/ 

```
make[2]: Leaving directory '/d/a/arrow/arrow/src/build-x86_64-cpp'
[ 48%] Built target arrow_testing_static
src/arrow/CMakeFiles/arrow-type-test.dir/build.make:84: *** target pattern contains no '%'.  Stop.
make[2]: Entering directory '/d/a/arrow/arrow/src/build-x86_64-cpp'
make[1]: *** [CMakeFiles/Makefile2:630: src/arrow/CMakeFiles/arrow-type-test.dir/all] Error 2
make[2]: Leaving directory '/d/a/arrow/arrow/src/build-x86_64-cpp'
make[1]: *** Waiting for unfinished jobs....
```

**Comment by kou** [comment]
We can use the gtest package provided by MSYS2 like https://github.com/apache/arrow/blob/master/ci/scripts/msys2_setup.sh#L35 instead of building bundled gtest. It'll reduce CI time.

**Comment by kou** [comment]
We're running C++ tests with MinGW 32bit/64bit: https://github.com/apache/arrow/blob/master/.github/workflows/cpp.yml#L301-L312

But it uses "release" build not "debug" build.

**Comment by kou** [comment]
For GLib & Ruby, we only use 64bit MinGW because https://github.com/actions/setup-ruby provides only 64bit Ruby.

File: ci/scripts/PKGBUILD [hunk 1]
```diff
@@ -50,6 +52,12 @@ source_dir="$ARROW_HOME"
 
 cpp_build_dir=build-${CARCH}-cpp
 
+# This should be "release" for real R package
+cpp_build_type="debug"
```
Comment (bkietz) [inline ci/scripts/PKGBUILD:56]: This seems like it should be derived from an environment variable near the top of this script:
```suggestion
cpp_build_type="${CPP_BUILD_TYPE:-debug}"
```

Then the value of `depends` can include boost iff we're building debug and this script can be used to build release by setting `CPP_BUILD_TYPE=release` or so

**Review by bkietz — CHANGES_REQUESTED** [review]
needs a rebase

**Comment by nealrichardson** [comment]
@github-actions rebase

**Comment by nealrichardson** [comment]
I'm closing this; don't think it's worth our time right now to pursue further. We can revisit later if we want.

**Pull request closed without merging**